src/kubernetes/Handlers/KubernetesResourceHandler.cs

```diff
@@ -1,46 +1,59 @@
 
     public async Task<LocalExtensibilityOperationResponse> Get(Protocol.ResourceReference request, CancellationToken cancellationToken)
-        => Convert(await new KubernetesProvider().GetAsync(Convert(request), cancellationToken), request.Type, request.ApiVersion);
+        => Convert(await GetKubernetesExtension().GetResourceAsync(new(), Convert(request), cancellationToken), request.Type, request.ApiVersion);
 
     public async Task<LocalExtensibilityOperationResponse> Preview(Protocol.ResourceSpecification request, CancellationToken cancellationToken)
-        => Convert(await new KubernetesProvider().PreviewSaveAsync(Convert(request), cancellationToken), request.Type, request.ApiVersion);
+        => Convert(await GetKubernetesExtension().PreviewResourceCreateOrUpdateAsync(new(), Convert(request), cancellationToken), request.Type, request.ApiVersion);
 
     public async Task<LocalExtensibilityOperationResponse> CreateOrUpdate(Protocol.ResourceSpecification request, CancellationToken cancellationToken)
-        => Convert(await new KubernetesProvider().SaveAsync(Convert(request), cancellationToken), request.Type, request.ApiVersion);
+        => Convert(await GetKubernetesExtension().CreateOrUpdateResourceAsync(new(), Convert(request), cancellationToken), request.Type, request.ApiVersion);
 
-    private static ExtCore.ExtensibilityOperationRequest Convert(Protocol.ResourceSpecification request)
-    {
-        return new(
-            new("Kubernetes", "1.0.0", JsonSerializer.Deserialize<JsonElement>(request.Config)),
-            new(request.ApiVersion is {} ? $"{request.Type}@{request.ApiVersion}" : request.Type, JsonSerializer.Deserialize<JsonElement>(request.Properties)));
-    }
+    private static V2Models.ResourceSpecification Convert(Protocol.ResourceSpecification request)
+        => new(
+            request.Type,
+            request.ApiVersion,
+            request.Properties,
+            request.Config);
 
-    private static ExtCore.ExtensibilityOperationRequest Convert(Protocol.ResourceReference request)
-    {
-        return new(
-            new("Kubernetes", "1.0.0", JsonSerializer.Deserialize<JsonElement>(request.Config)),
-            new(request.ApiVersion is {} ? $"{request.Type}@{request.ApiVersion}" : request.Type, JsonSerializer.Deserialize<JsonElement>(request.Identifiers)));
-    }
+    private static V2Models.ResourceReference Convert(Protocol.ResourceReference request)
+        => new(
+            request.Type,
+            request.ApiVersion,
+            request.Identifiers,
+            request.Config);
 
-    private static Protocol.LocalExtensibilityOperationResponse Convert(ExtCore.ExtensibilityOperationResponse response, string type, string? apiVersion)
+    private static Protocol.LocalExtensibilityOperationResponse Convert(IResult response, string type, string? apiVersion)
```

Comment:
How to convert `IResult` back to `Protocol.LocalExtensibilityOperationResponse`?

Reply:
```C#
var resource = response is Ok<Resource> ok ? ok.Value : null;
var errorData = response switch
    {
        BadRequest<ErrorData> badRequest => badRequest.Value,
        NotFound<ErrorData> notFound => notFound.Value,
        // The k8s extension only returns Ok, BadRequest, and NotFound.
        _ => null,
    };

var converted = new Protocol.LocalExtensibilityOperationResponse(resource, errorData);
```